File: driver-api/src/main/scala/net/cakesolutions/reactiveredis/driver/api/commands.scala
package net.cakesolutions.reactiveredis.driver.api

import scala.collection.JavaConverters._
import scala.reflect.ClassTag

object commands {
  sealed trait RedisRequest

  trait KeyCheck
  case object KeyMustNotExist extends KeyCheck
  case object KeyMustExist extends KeyCheck
  case object NoKeyCheck extends KeyCheck

  sealed trait RedisCommand extends RedisRequest
  case object Dbsize extends RedisCommand
  case object Flushall extends RedisCommand
  case class Get[R: ClassTag](key: String) extends RedisCommand {
    val returnClassTag = scala.reflect.classTag[R]
  }
  case class Hlen(key: String) extends RedisCommand
  case class Hmset(key: String, map: Map[String, String]) extends RedisCommand
  case object Ping extends RedisCommand
  case class Set(key: String, value: String, ex: Option[Long], px: Option[Long], keyCheck: KeyCheck) extends RedisCommand

  case class Pipeline(commands: List[RedisCommand]) extends RedisRequest

  def keyMustNotExist = KeyMustNotExist
  def keyMustExist = KeyMustExist
  def noKeyCheck = NoKeyCheck

  def dbsize = Dbsize
  def flushall = Flushall
  def get[R: ClassTag](key: String) = Get[R](key)
  def hlen(key: String) = Hlen(key)
  def hmset(key: String, map: Map[String, String]) = Hmset(key, map)
  def ping = Ping
  def set(key: String, value: String, ex: Option[Long], px: Option[Long], keyCheck: KeyCheck) = Set(key, value, ex, px, keyCheck)
  def set(key: String, value: String) = Set(key, value, None, None, noKeyCheck)

  def jhmset(key: String, map: java.util.Map[String, String]) = hmset(key, map.asScala.toMap)

  def pipeline(commands: List[RedisCommand]) = Pipeline(commands)
  def jpipeline(commands: java.util.List[RedisCommand]) = pipeline(commands.asScala.toList)
}

File: driver-api/src/main/java/net/cakesolutions/reactiveredis/driver/japi/Commands.java
package net.cakesolutions.reactiveredis.driver.japi;

import net.cakesolutions.reactiveredis.driver.api.commands;
import net.cakesolutions.reactiveredis.driver.api.commands.RedisCommand;
import scala.Option;
import scala.reflect.ClassTag$;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.OptionalLong;

public class Commands {
    public static commands.KeyCheck keyMustNotExist() {
        return commands.keyMustNotExist();
    }

    public static commands.KeyCheck keyMustExist() {
        return commands.keyMustExist();
    }

    public static commands.KeyCheck noKeyCheck() {
        return commands.noKeyCheck();
    }

    public static RedisCommand dbsize() {
        return commands.dbsize();
    }

    public static RedisCommand flushall() {
        return commands.flushall();
    }

    public static <R> RedisCommand get(String key, Class<R> clazz) {
        return commands.<R>get(key, ClassTag$.MODULE$.apply(clazz));
    }

    public static RedisCommand hlen(String key, Map<String, String> map) {
        return commands.jhmset(key, map);
    }

    public static RedisCommand hmset(String key, Map<String, String> map) {
        return commands.jhmset(key, map);
    }

    public static RedisCommand ping() {
        return commands.ping();
    }

    public static RedisCommand set(String key, String value) {
        return commands.set(key, value);
    }

    public static RedisCommand set(String key, String value, OptionalLong ex, OptionalLong px, commands.KeyCheck keyCheck) {
        return commands.set(key, value, toScalaLong(ex), toScalaLong(px), keyCheck);
    }

    public static commands.RedisRequest pipeline(List<RedisCommand> redisCommands) {
        return commands.jpipeline(redisCommands);
    }

    private static Option<Object> toScalaLong(OptionalLong optional) {
        return optional.isPresent() ? Option.apply(optional.getAsLong()) : Option.empty();
    }

    private static <T> Option<T> toScala(Optional<T> optional) {
        return optional.isPresent() ? Option.apply(optional.get()) : Option.empty();
    }
}
